pkg/virt-launcher/virtwrap/manager.go

```diff
@@ -490,6 +493,46 @@ monitorLoop:
 	}
 }
 
+func (l *LibvirtDomainManager) CancelVMIMigration(vmi *v1.VirtualMachineInstance) error {
+	if vmi.Status.MigrationState == nil || vmi.Status.MigrationState.Completed ||
+		vmi.Status.MigrationState.Failed {
+
+		l.setMigrationResult(vmi, false, "", v1.MigrationAbortFailed)
+		return fmt.Errorf("failed to cancel migration - vmi is not migrating")
+	}
+	l.asyncMigrationAbort(vmi)
```

Comment:
What keeps us from spawning multiple goroutines here if CancelVMIMigraiton is called multiple times.  It's possible for that to occur. That's why MigrateVMI has to ensure there isn't already a migration in progress before calling asyncMigrate.

Reply:
hmm... but it does in initializeMigrationMetadata()

Reply:
initializeMigrationmetadata prevents multiple migrations from occuring.

I don't see anything preventing multiple asyncMigrationAbort()  goroutines from spawning though. For example, in MigrateVMI we check to see if the migraiton is in-progress before migrating. We aren't checking to see if the abort migration is in progress here. Maybe there's something I'm missing?

Reply:
ah, ok. I think I misunderstood your previous comment.

Reply:
@davidvossel do you think we should add another field to the domain metadata or a local file will do?
Unlike in the migration case, here there is nothing in libvirt that indicates that the migration is being aborted.

On the other hand, I can't think of what harm can be done if two cancel calls been made. The second one will always fail.

Reply:
> do you think we should add another field to the domain metadata or a local file will do?

the domain's metadata could be used again. The trick would be that the metadata would need to be updated if the job failed to ensure that the job abort is allowed to be attempted again. 

> On the other hand, I can't think of what harm can be done if two cancel calls been made. The second one will always fail.

i'm really not sure what the ramifications are here. Depending on the time involved, it could be a lot more than 2 go-routines. It's really unbounded. Every time the vmi is synced, a new cancel go-routine will get spawned while the job is being aborted.